server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKSecurityTool.java

```diff
@@ -114,16 +120,27 @@ public static byte[] createPass(byte[] password) throws AccumuloException {
     return cryptHash.getBytes(UTF_8);
   }
 
+  private static final Map<List<Byte>,String> CHECKED_CRYPT_PASSWORDS =
+      Collections.synchronizedMap(new LRUMap<>(16));
+
   public static boolean checkCryptPass(byte[] password, byte[] zkData) {
+    List<Byte> key = Bytes.asList(password);
     String zkDataString = new String(zkData, UTF_8);
+    if (CHECKED_CRYPT_PASSWORDS.getOrDefault(key, "").equals(zkDataString)) {
+      return true;
+    }
     String cryptHash;
     try {
       cryptHash = Crypt.crypt(password, zkDataString);
```

Comment:
Feels more safe to me if we explicitly cache the input and output of the cryptHash function. So something like the followin psuedocode (not sure of the best way to write it w/o looking at cache APIs)

```
   var key = password  + zkData
   cryptHash = getFromCache(key)
   if(cryptHash == null) {
       cryptHash = Crypt.crypt(password, zkDataString);
       putInCache(key, cryptHash);
   }
   
   boolean matches = MessageDigest.isEqual(zkData, cryptHash.getBytes(UTF_8));
```

Reply:
I was thinking that we would not want to store the output of any crypt function. Instead, I was just capturing the fact that two pieces of information had returned a true result in the recent past, and using it to short circuit the method. Is there any benefit to recalculating whether the two hashes are equal?

Reply:
Was just thinking very literally and trying to avoid shortcuts since its security related.  The function takes two inputs and gives a single output,  so was thinking of caching that functions input and output exactly rather than caching a partial was more secure.  Caching only a single input parameter (the password) felt kind dicey.  Like what if two account have the same password with different salts, then what will happen with the cache?

Reply:
> Ah, I see. I didn't think about the password reuse case. I'll adjust.